### PR TITLE
Add Flowstar (flowstar.org)

### PR DESCRIPTION
**Pull request by agerlach**
Windows not currently supported.

### CODE REVIEW

File: F/Flowstar/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+using BinaryBuilder
+
+platforms = supported_platforms(exclude=Sys.iswindows)
+platforms = expand_cxxstring_abis(platforms)
+
+name = "Flowstar"
+version = v"2.1.0"
+sources = [
+    ArchiveSource("https://www.cs.colorado.edu/~xich8622/src/flowstar-$version.tar.gz", "642b17a55c6725d4bfc5b98900802e3b82f37fbbe9fb9028f1110c669a5afc86")
+    FileSource("https://www.gnu.org/licenses/gpl-3.0.txt", "3972dc9744f6499f0f9b2dbf76696f2ae7ad8af9b23dde66d6af86c9dfb36986"; filename = "LICENSE")
```
Comment (giordano) [inline F/Flowstar/build_tarballs.jl:16]: We already ship the GPL license in the build environment, no need to download it
```suggestion
```

File: F/Flowstar/build_tarballs.jl [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+begin
+    script = raw"""
+    cd ${WORKSPACE}/srcdir/flowstar-VERSION
+    make -j${nproc}
+    chmod +x flowstar
+    cp flowstar ${bindir}
+    """
+    script = replace(script,"VERSION"=>"$version")
+end
```
Comment (giordano) [inline F/Flowstar/build_tarballs.jl:27]:
```suggestion
script = raw"""
cd ${WORKSPACE}/srcdir/flowstar*
make -j${nproc}
install -Dvm 0755 flowstar "${bindir}/flowstar${exeext}"
install_license /usr/share/licenses/GPL-3.0+
"""
```

File: F/Flowstar/build_tarballs.jl [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    HostBuildDependency("flex_jll"),
+    HostBuildDependency("Bison_jll"),
+    Dependency("MPFR_jll"),
+    Dependency("GSL_jll"),
+    Dependency("GLPK_jll"),
+    Dependency("GMP_jll")
+]
+
+res = build_tarballs(ARGS, name, version, sources, script, platforms, products, dependencies; 
+        julia_compat = "1.6", preferred_gcc_version=v"8")
```
Comment (giordano) [inline F/Flowstar/build_tarballs.jl:43]:
```suggestion
        julia_compat = "1.6", preferred_gcc_version=v"8")
```

File: F/Flowstar/build_tarballs.jl [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+products = [
+    ExecutableProduct("flowstar", :flowstar)
+]
+
+dependencies = [
+    HostBuildDependency("flex_jll"),
+    HostBuildDependency("Bison_jll"),
+    Dependency("MPFR_jll"),
+    Dependency("GSL_jll"),
```
Comment (giordano) [inline F/Flowstar/build_tarballs.jl:37]:
```suggestion
    Dependency("GSL_jll"; compat="~2.7.2"),
```

**Review by giordano — COMMENTED** [review]
What's the problem on Windows?

File: F/Flowstar/build_tarballs.jl (hunk 1 above)
Comment (agerlach) [reply F/Flowstar/build_tarballs.jl:16]: If I don't include I LICENSE file I don't pass the audit. Do I need to copy it from somewhere?

Comment (agerlach) [reply F/Flowstar/build_tarballs.jl:16]: Oh... I see it in the script change below.

Comment (giordano) [reply F/Flowstar/build_tarballs.jl:16]: That's what the line `install_license /usr/share/licenses/GPL-3.0+` I suggested adding in https://github.com/JuliaPackaging/Yggdrasil/pull/5035#discussion_r900024955 is for

File: F/Flowstar/build_tarballs.jl (hunk 3 above)
Comment (agerlach) [reply F/Flowstar/build_tarballs.jl:39]: There is no change here. Am I missing something?

Comment (giordano) [reply F/Flowstar/build_tarballs.jl:39]: The newline I added disappeared, now it's back.  The file is missing the final newline

**Comment by agerlach** [comment]
> What's the problem on Windows?

I wasn't able to get it to compile. 

```
Continuous.cpp: In member function ‘void flowstar::ContinuousReachability::dump(FILE*) const’:
Continuous.cpp:14152:86: error: too many arguments to function ‘int mkdir(const char*)’
  int mkres = mkdir(counterexampleDir, S_IRWXU | S_IRGRP | S_IXGRP | S_IROTH | S_IXOTH);
                                                                                      ^
In file included from /opt/i686-w64-mingw32/i686-w64-mingw32/sys-root/include/sys/stat.h:14,
                 from include.h:15,
                 from Interval.h:12,
                 from Monomial.h:12,
                 from Polynomial.h:12,
                 from TaylorModel.h:12,
                 from Continuous.h:12,
                 from Continuous.cpp:9:
/opt/i686-w64-mingw32/i686-w64-mingw32/sys-root/include/io.h:280:15: note: declared here
   int __cdecl mkdir (const char *) __MINGW_ATTRIB_DEPRECATED_MSVC2005;
               ^~~~~
Continuous.cpp: In member function ‘int flowstar::ContinuousReachability::safetyChecking()’:
Continuous.cpp:14848:87: error: too many arguments to function ‘int mkdir(const char*)’
   int mkres = mkdir(counterexampleDir, S_IRWXU | S_IRGRP | S_IXGRP | S_IROTH | S_IXOTH);
                                                                                       ^
In file included from /opt/i686-w64-mingw32/i686-w64-mingw32/sys-root/include/sys/stat.h:14,
                 from include.h:15,
                 from Interval.h:12,
                 from Monomial.h:12,
                 from Polynomial.h:12,
                 from TaylorModel.h:12,
                 from Continuous.h:12,
                 from Continuous.cpp:9:
/opt/i686-w64-mingw32/i686-w64-mingw32/sys-root/include/io.h:280:15: note: declared here
   int __cdecl mkdir (const char *) __MINGW_ATTRIB_DEPRECATED_MSVC2005;
               ^~~~~
```

I don't know of anybody that has actually compile Flow* for windows successfully.

**Comment by giordano** [comment]
You need something like https://github.com/JuliaPackaging/Yggdrasil/blob/d936efb0feaa2665c1b25766d2a1bc2258738d6d/F/Fontconfig/bundled/patches/0002-fix-mkdir.mingw.patch

File: F/Flowstar/build_tarballs.jl [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+name = "Flowstar"
+version = v"2.1.0"
+sources = [
+    ArchiveSource("https://www.cs.colorado.edu/~xich8622/src/flowstar-$version.tar.gz", "642b17a55c6725d4bfc5b98900802e3b82f37fbbe9fb9028f1110c669a5afc86")
+    DirectorySource("./bundled")
+]
+
+script = raw"""
+cd ${WORKSPACE}/srcdir/flowstar-2.1.0
+atomic_patch -p1 "${WORKSPACE}/srcdir/patches/windows_mkdir.patch"
```
Comment (giordano) [inline F/Flowstar/build_tarballs.jl:21]: The file is called `window_mkdir.patch`, not `windows_mkdir.patch`

File: F/Flowstar/bundled/patches/windows_mkdir.patch [hunk 6]
```diff
@@ -0,0 +1,15 @@
+diff --git a/include.h b/include.h
+index e246927..ff3e8ad 100644
+--- a/include.h
++++ b/include.h
+@@ -31,6 +31,10 @@
+ #include <gsl/gsl_linalg.h>
+ #include <glpk.h>
+ 
++#ifdef _WIN32
```
Comment (giordano) [inline F/Flowstar/bundled/patches/windows_mkdir.patch:9]: `_mkdir` requires `direct.h`: https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/mkdir-wmkdir?view=msvc-170#requirements
```suggestion
+++ b/include.h
@@ -31,6 +31,11 @@
 #include <gsl/gsl_linalg.h>
 #include <glpk.h>
 
+#ifdef _WIN32
#include <direct.h>
```

File: F/Flowstar/bundled/patches/windows_mkdir.patch [hunk 7]
```diff
@@ -0,0 +1,16 @@
+diff --git a/include.h b/include.h
+index e246927..ff3e8ad 100644
+--- a/include.h
++++ b/include.h
+@@ -31,6 +31,11 @@
+ #include <gsl/gsl_linalg.h>
+ #include <glpk.h>
+ 
++#ifdef _WIN32
+#include <direct.h>
```
Comment (giordano) [inline F/Flowstar/bundled/patches/windows_mkdir.patch:10]: Remind me not to manually edit patch files :roll_eyes: 
```suggestion
+#include <direct.h>
```

**Comment by giordano** [comment]
```
[14:56:31] g++ -O3 -c -I . -I /usr/local/include -g -O3 -std=c++11 -o lex.yy.o lex.yy.c
[14:56:53] g++ -O3 -w -g -L/usr/local/lib  -o flowstar Interval.o Matrix.o Polynomial.o Monomial.o Constraints.o Geometry.o TaylorModel.o Continuous.o Hybrid.o lex.yy.o modelParser.tab.o modelParser.o -lmpfr -lgmp -lgsl -lgslcblas -lm -lglpk
```
two errors in two lines, great....  I'm going to try and patch this stuff out.

File: F/Flowstar/bundled/patches/windows_mkdir.patch (hunk 7 above)
Comment (agerlach) [reply F/Flowstar/bundled/patches/windows_mkdir.patch]: > Remind me not to manually edit patch files 🙄

😆  Does that earn a YOLO badge? 

Thanks for all you help and patience. I'm a bit out of my element here

**Comment by agerlach** [comment]
I guess that means the exe doesn't exist. It's weird that their is no compilation error though

**Comment by giordano** [comment]
Thanks for your contribution!

**Comment by agerlach** [comment]
Thanks again... now the hard work of tying it into ReachabilityAnalylsis.jl

**Comment by agerlach** [comment]
How long does it typically take before showing up in the general registry?

**Pull request merged**